### PR TITLE
Add sts_session_token module for short-lived AWS credentials.

### PR DESCRIPTION
**Pull request by pwnall**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

sts_session_token
##### ANSIBLE VERSION

```
ansible 2.0.1.0
```
##### SUMMARY

This module wraps the [GetSessionToken API](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html), similarly to how the [sts_assume_role module](http://docs.ansible.com/ansible/sts_assume_role_module.html) wraps the [AssumeRole API](http://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html).

This module's most likely use is obtaining API credentials for accounts where Multi-Factor Authentication (MFA) is required.

I tested my code with the following rules:

```
- name: obtain MFA code
  sts_session_token:
    region: "{{ aws_region }}"
    duration_seconds: 3600
    mfa_serial_number: "{{ mfa_serial_number }}"
    mfa_token: "{{ mfa_token }}"
  register: session_credentials

- debug: var=session_credentials
```

Output:

```
ok: [localhost] => {
    "session_credentials": {
        "changed": true,
        "sts_creds": {
            "access_key": "ASXXXXXXXXXXXXXXXXXX",
            "expiration": "2016-04-08T11:59:47+00:00",
            "secret_key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
            "session_token": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        }
    }
}
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @pwnall for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by dorianj** [comment]
shipit -- I've been using this locally and it works for me.  (along with https://github.com/ansible/ansible-modules-extras/pull/1987)

**Comment by gregdek** [comment]
Thanks @pwnall, and thanks @dorianj for the review -- will get into the final review pipeline.

**Comment by gregdek** [comment]
Thanks @pwnall for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by pwnall** [comment]
@gregdek PTAL?

(Edit by @gregdek -- adding text ready_for_review so bot picks it up.)

**Comment by gregdek** [comment]
Thanks @pwnall for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by pwnall** [comment]
@gregdek This PR already has a [shipit endorsement](https://github.com/ansible/ansible-modules-extras/pull/1988#issuecomment-212208059). The only change I had to make since then involved bumping `version_added`.

File: cloud/amazon/sts_session_token.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        module.fail_json(msg="region must be specified")
+
+    get_session_token(connection, module)
+
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
+
+main()
```
Comment (gundalow) [inline cloud/amazon/sts_session_token.py]: This needs to be as follows so debuggers can work with ziploader:

```
if __name__ == '__main__':
    main()
```

**Comment by gundalow** [comment]
needs_revision 

Nothing major. The previous endorcements should still count once my comments have been addressed.

**Comment by pwnall** [comment]
@gundalow Can you please take another look? I hope I've addressed your feedback.

File: cloud/amazon/sts_session_token.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      - The value provided by the MFA device, if the trust policy of the user requires MFA.
+    required: false
+    default: null
+notes:
+  - In order to use the session token in a following playbook task you must pass the access_key, access_secret and access_token
+extends_documentation_fragment:
+    - aws
+    - ec2
+requirements:
+    - boto3
```
Comment (gundalow) [inline cloud/amazon/sts_session_token.py:47]: - botocore

File: cloud/amazon/sts_session_token.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            duration_seconds = dict(required=False, default=None, type='int'),
+            mfa_serial_number = dict(required=False, default=None),
+            mfa_token = dict(required=False, default=None)
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+      module.fail_json(msg='boto3 is required.')
```
Comment (gundalow) [inline cloud/amazon/sts_session_token.py]: `boto3 and botocore are required.`

Also the indentation looks like it's missing two spaces.

**Comment by gundalow** [comment]
I wonder if this should be called `aws_sts_session_token.py`

**Comment by gundalow** [comment]
needs_revision

**Comment by ryansb** [comment]
@gundalow IMO `aws_sts_token` would be better, since the "session" bit is implied by being an STS resource.

**Comment by pwnall** [comment]
@ryansb @gundalow Thank you very much for your feedback! I changed the name to `aws_sts_token`. Here are a few things I had in mind when I chose the previous name, `sts_session_token`:
- Ansible has an [sts_assume_role module](http://docs.ansible.com/ansible/sts_assume_role_module.html), so I assumed that the `sts_` prefix has been dedicated to the AWS STS.
- [sts_assume_role module](http://docs.ansible.com/ansible/sts_assume_role_module.html) wraps the [AssumeRole API](http://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html), and this module wraps the [GetSessionToken API](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html)
- The Secure Token Service (STS) also has an [GetFederationToken API](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetFederationToken.html) that returns a different kind of token

Please let me know if you'd like me to make any changes to address the inconsistency between `sts_assume_role` and this module's naming scheme.

**Comment by ryansb** [comment]
@pwnall very fair point, looking at the existing STS modules, it's better to go with your original name (I also wasn't aware of STS federation tokens, which throw in a wrinkle).

With profuse apologies for the runaround, sts_session_token seems to be the right name.

**Comment by pwnall** [comment]
@ryansb Thank you very much for the feedback! Ansible has made my life better, and I'm very happy to be able to contribute back! Please let me know if there's anything else I can do to help this get merged.

File: cloud/amazon/sts_session_token.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+module: sts_session_token
+short_description: Obtain a session token from the AWS Security Token Service
+description:
+    - Obtain a session token from the AWS Security Token Service
+version_added: "2.2"
+author: Victor Costan (@pwnall)
+options:
+  duration_seconds:
+    description:
+      - The duration, in seconds, of the session token. The value can range from 900 seconds (15 minutes) to 129600 seconds (36 hours). By default, the value is set to 3600 seconds (1 hour) for an AWS account or 43200 seconds for an IAM account.
```
Comment (gundalow) [inline cloud/amazon/sts_session_token.py]: Are the defaults hard coded in AWS itself?

Could they ever change and the `description` be out of sync with the code?

I can't think of a better way of addressing this, just asking the question

File: cloud/amazon/sts_session_token.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      - The identification number of the MFA device that is associated with the user who is making the GetSessionToken call.
+    required: false
+    default: null
+  mfa_token:
+    description:
+      - The value provided by the MFA device, if the trust policy of the user requires MFA.
+    required: false
+    default: null
+notes:
+  - In order to use the session token in a following playbook task you must pass the access_key, access_secret and access_token
```
Comment (gundalow) [inline cloud/amazon/sts_session_token.py]: We can make this a little more readable by adding formatting

From http://docs.ansible.com/ansible/developing_modules.html#module-checklist
- Ensure that you make use of U() for urls, C() for files and options, I() for params, M() for modules.

`- In order to use the session token in a following playbook task you must pass the I(access_key), I(access_secret) and I(access_token).
`

**Comment by gundalow** [comment]
Hi @pwnall thank you for this new module, and your understanding about trying to get the name right. Once the module is in it's a pain to rename stuff, so we like to think about this stuff up front, unfortunately that sometimes means a bit of back and forth.

I've completed my final review, just one question (which I don't think we can do anything about) and one small formatting issue. Once those are addressed you will get my stamp of approval

needs_revision

**Comment by pwnall** [comment]
@gundalow Thank you very much for pointing out the formatting issue! I'm really grateful for your help in making this module more usable!

The defaults and limits for the duration are indeed wired into AWS. We can't do anything about the limits, but we could safeguard against default value changes by setting our own default. Without extra API calls, we can't distinguish between AWS and IAM accounts, so we'd have to set the default to 3600, which is a safe value for both account types.

Please let me know if you'd like me to set a default and update the documentation.

**Comment by gundalow** [comment]
Thank you for quickly addressing those issues. 

Regarding the defaults thanks for having more of a look into this. I think having the same defaults of 3600 could be confusing as it's different to it the API is called directly, or setup via the AWS console. 

Then again I'm not sure if it's worth an an extra API call just to find the defaults out. 

Maybe just add a note to say that the default values are set in AWS, and a link to the page(s) that define them.

**Comment by gregdek** [comment]
Thanks @pwnall for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by pwnall** [comment]
@gundalow What do you think of this? `See http://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html#API_GetSessionToken_RequestParameters for acceptable and default values.`

@gregdek ready_for_review

**Comment by gregdek** [comment]
Thanks @pwnall for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gundalow** [comment]
@pwnall That link to docs.aws.amazon.com looks spot on, thanks.

meets_guidelines

**Comment by gundalow** [comment]
shipit

**Comment by abadger** [comment]
Merged to devel.

**Comment by pwnall** [comment]
@gundalow @ryansb Thank you very much for your help in shipping this module! ❤️

**Comment by gundalow** [comment]
@pwnall Thanks for getting all the issues addressed promptly

**Pull request merged**